enaml/core/code_tracing.py

```diff
@@ -345,13 +353,31 @@ def inject_tracing(codelist):
         elif op == RETURN_VALUE:
             code = [
                 (DUP_TOP, None),                # obj
-                (LOAD_FAST, '_[tracer]'),       # obj -> obj -> tracer
+                (tracer_op, '_[tracer]'),       # obj -> obj -> tracer
                 (LOAD_ATTR, 'return_value'),    # obj -> obj -> tracefunc
                 (ROT_TWO, None),                # obj -> tracefunc -> obj
                 (CALL_FUNCTION, 0x0001),        # obj -> retval
                 (POP_TOP, None),                # obj
             ]
             inserts[idx] = code
+        elif isinstance(op_arg, bp.Code):
+            # Inject tracing in nested code object.
+            # This handles the case of list/dict/set comprehensions that
+            # defines a nested function.
+
+            # As we cannot pass the tracer as argument to the function, get it
+            # from the dynamicscope that will find it in the fast locals and
+            # cache it. This will allow to access to it later
```

Comment:
I  don't believe that DynamicScope caches lookups. If the code below `locals()` is returning the dynamic scope, you can instead just do `(LOAD_NAME, '__scope__')` which is special-cased by the dynamic scope to return itself. That said, I'm not sure where the dynamic scope is ever given a handle to a tracer in a way where it is "on the lookup path".

I'm also not sure that the nested functions are going to be called with the same `f_locals` object (which is the dynamic scope) without special work to set up the call.

Reply:
Ah. It's the actual call to `locals()` which does the fetching of fast locals and storing them in slow locals. The internal call which does the work is `PyFrame_FastToLocalsWithError`.

So this approach seems to "technically work", except for understand how the dynamic scope gets passed to the nested function as the `f_locals`.

Reply:
This is handled as part of the support for comprehension. See compiler_utils for this part.

Reply:
Ok. I'm still not a fan of how this is accomplished. I would still either rewrite the inner comprehension code to take `_[tracer]` as part of fast locals, *or*, update `DynamicScope` to make `_[tracer]` available as a "magic". Then you won't need this `locals()` dance.

Reply:
Do you have a preference between the two options (just to know before starting working on this again) ? I think the fix on the dynamic scope will be easier and probably look better.

Reply:
The dynamic scope will definitely be the easier fix. But the `fast_locals` approach would make things more "uniform". I'll let you make the call.